File: ressources/[Exemple] Episodic Semi gradient SARSA on TicTacToeVersus Random in Rust _ Burn/src/contracts.rs
use burn::prelude::{Backend, Tensor};

pub trait Forward {
    type B: Backend;
    fn forward<const DIM: usize>(&self, input: Tensor<Self::B, DIM>) -> Tensor<Self::B, DIM>;
}

pub trait DeepDiscreteActionsEnv<const NUM_STATES_FEATURES: usize, const NUM_ACTIONS: usize>: Default + Clone {
    fn state_description(&self) -> [f32; NUM_STATES_FEATURES];
    fn available_actions_ids(&self) -> impl Iterator<Item=usize>;
    fn action_mask(&self) -> [f32; NUM_ACTIONS];
    fn step(&mut self, action: usize);
    fn is_game_over(&self) -> bool;
    fn score(&self) -> f32;
    fn reset(&mut self);
}

File: ressources/[Exemple] Episodic Semi gradient SARSA on TicTacToeVersus Random in Keras_TF_Keras_PyTorch_PyTorch/contracts.py
import tensorflow as tf
import numpy as np

class Forward:
    def forward(self, input_tensor: tf.Tensor) -> tf.Tensor:
        raise "Unimplemented"

class DeepDiscreteActionsEnv:
    def state_description(self) -> np.ndarray:
        raise "Unimplemented"

    def available_actions_ids(self) -> np.ndarray:
        raise "Unimplemented"

    def action_mask(self) -> np.ndarray:
        raise "Unimplemented"

    def step(self, action: int):
        raise "Unimplemented"

    def is_game_over(self) -> bool:
        raise "Unimplemented"

    def score(self) -> float:
        raise "Unimplemented"

    def reset(self):
        raise "Unimplemented"